apps/arweave/c_src/ar_mine_randomx.c

```diff
@@ -337,6 +353,106 @@ static ERL_NIF_TERM encrypt_chunk(ErlNifEnv* envPtr,
 	return encryptedChunkTerm;
 }
 
+static ERL_NIF_TERM encrypt_composite_chunk(ErlNifEnv* envPtr,
+		randomx_vm *vmPtr, ErlNifBinary *inputDataPtr, ErlNifBinary *inputChunkPtr,
+		const int subChunkCount, const int iterations,
+		const int randomxRoundCount, const int jitEnabled,
+		const int largePagesEnabled, const int hardwareAESEnabled) {
+
+	unsigned char *paddedChunk = (unsigned char*)malloc(MAX_CHUNK_SIZE);
+	if (inputChunkPtr->size == MAX_CHUNK_SIZE) {
+		memcpy(paddedChunk, inputChunkPtr->data, inputChunkPtr->size);
+	} else {
+		memset(paddedChunk, 0, MAX_CHUNK_SIZE);
+		memcpy(paddedChunk, inputChunkPtr->data, inputChunkPtr->size);
+	}
+
+	ERL_NIF_TERM encryptedChunkTerm;
+	unsigned char* encryptedChunk = enif_make_new_binary(envPtr, MAX_CHUNK_SIZE,
+			&encryptedChunkTerm);
+	// Both MAX_CHUNK_SIZE and subChunkCount are multiples of 64 so all sub-chunks
+	// are of the same size.
+	uint32_t subChunkSize = MAX_CHUNK_SIZE / subChunkCount;
+	uint32_t offset = 0;
+	unsigned char key[PACKING_KEY_SIZE];
+	for (int i = 0; i < subChunkCount; i++) {
+		unsigned char* subChunk = paddedChunk + offset;
+		unsigned char* encryptedSubChunk = (unsigned char*)malloc(subChunkSize);
+
+		// 3 bytes is sufficient to represent offsets up to at most MAX_CHUNK_SIZE.
+		int offsetByteSize = 3;
+		unsigned char offsetBytes[offsetByteSize];
+		for (int k = 0; k < offsetByteSize; k++) {
+			offsetBytes[k] = ((offset + subChunkSize) >> (8 * (offsetByteSize - 1 - k))) & 0xFF;
+		}
+		SHA256_CTX sha256;
```

Comment:
Is this a correct understanding of this block?
1. It computes a feistel hash which is used as the encryption key for the sub-chunk
2. the feistel hash is generated from the packing address (i.e. `inputData`) and the offset of the sub-chunk
3. this guarantees that the encryption key for each sub-chunk is unique but deterministic

If that's correct maybe we could add something like that as a comment? e.g.

```
// Sub-chunk encryption key is the feistel hash of the input data and the sub-chunk offset
```

Reply:
It is a sha2 hash (not a feistel)

Reply:
haha, my ignorance is showing. I was going off of pattern matching with the `feistel_hash()` function: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/c_src/feistel_msgsize_key_cipher.cpp#L7-L13

The code blocks look pretty similar - is the main difference that a feistel hash forces a 32-length? (or is it just a misnaming of the `feistel_hash(...)` function and I should ignore?)

Reply:
The block computes sha256(<< packing_key_without_sub_chunk_offset, sub_chunk_offset >>).